### PR TITLE
Fix ON GOTO and ON GOSUB - Should resolve #11

### LINKED ISSUE #11: ON expr GOTO should proceed to next statement if expr is 0 or greater than the number of listed linenumbers.
ON expr GOTO linenum \[, ...\] (and ON GOSUB) should do nothing if expr is 0 or greater than the number of listed linenumbers (program execution proceeds to the next statement).
Unless expr is greater than 255, then we have a ILLEGAL QUANTITY ERROR.

See quote from http://mirrors.apple2.org.za/Apple%20II%20Documentation%20Project/Software/Languages/Applesoft%20BASIC/Manuals/Applesoft%20II%20BASIC%20Programming%20Reference%20Manual.pdf (page 81):

> ON aexpr GOTO linenum {[, linenum]}
> ON aexpr GOSUB linenum {[, linenum]}
> 
> ON...GOTO branches to the line number specified by the \aexpr\th item in the
> list of linenums after the GOTO. ON...GOSUB works in a similar fashion, but
> a GOSUB rather than a GOTO is executed.
> 
> If \aexpr\ is 0 or greater than the number of listed alternate linenums but
> less than 256, then program execution proceeds to the next statement.
> 
> \aexpr\ must be in the range 0 to 255 to avoid the message
> ?ILLEGAL QUANTITY ERROR

### PR DESCRIPTION
**Pull request by siggisv**

### CODE REVIEW

File: basic.js [hunk 1]
```diff
@@ -523,13 +523,15 @@ this.basic = (function() {
       },
 
       'on_goto': function ON_GOTO(index /* , ...lines */) {
+        if (index < 0 || index > 255) {
```
Comment (inexorabletash) [inline basic.js]: This doesn't handle index === 255.1
I think we want `index = Math.floor(index)`

File: basic.js [hunk 2]
```diff
@@ -523,13 +523,15 @@ this.basic = (function() {
...
+          runtime_error(ERRORS.ILLEGAL_QUANTITY);
+        }
         index = (index - 1) >> 0;
```
Comment (inexorabletash) [inline basic.js]: If we do `index = Math.floor(index)` above then this can just be `--index;`

File: basic.js [hunk 3]
```diff
@@ -541,16 +543,18 @@ this.basic = (function() {
       },
 
       'on_gosub': function ON_GOSUB(index /* , ...lines */) {
+        if (index < 0 || index > 255) {
```
Comment (inexorabletash) [inline basic.js]: Same notes as for GOTO

**Comment by inexorabletash** [comment]
Thanks for the bug/fix! I left a couple of line notes after testing edge cases like -0.1, 255.1, etc on a real (well, emulated) Apple. 

Also, can you add tests to https://github.com/inexorabletash/jsbasic/blob/master/samples/sample.unittests.txt ?

**Comment by siggisv** [comment]
Ok. I'll make changes to handle the edge cases better and try if I can add tests to the unit tests (my JavaScript is a bit rusty) :)

**Comment by siggisv** [comment]
First time ever I've written unit tests in Applesoft BASIC.
If I understood your code right, then that should test all the cases we need.

**Comment by siggisv** [comment]
Figured out when I woke up this morning that in the unit tests I added yesterday, I forgot that the ONERR GOTO that I used needed to be turned off after use. I also figured that I needed to add more ONERR GOTO to catch the errors that might come up if some of my tests would fail.

**Comment by inexorabletash** [comment]
Yeah, the test harness is a bit... tricky. :)

I'll have a chance to try out the change and review in detail in a day or two. Sorry about the delay, and thank you *very* much for the bug/fix/tests!

**Comment by inexorabletash** [comment]
Patch looked great. Thanks again!

**Comment by siggisv** [comment]
You're welcome. :smile:
